jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -308,59 +324,153 @@ private class Watches private(
-    val trainMatrix = new DMatrix(trainIter1, cacheDirName.map(_ + "/train").orNull)
-    val trainGroups = buildGroups(trainIter2.map(_.group).toSeq).toArray
-    trainMatrix.setGroup(trainGroups)
+    val trainPoints = trainLabelPointGroups.flatMap(_.iterator)
+    val trainMatrix = new DMatrix(trainPoints, cacheDirName.map(_ + "/train").orNull)
+    trainMatrix.setGroup(trainGroups.result())
 
-    val testMatrix = new DMatrix(testPoints.iterator, cacheDirName.map(_ + "/test").orNull)
+    val testMatrix = new DMatrix(testPoints.result().iterator, cacheDirName.map(_ + "/test").orNull)
     if (trainTestRatio < 1.0) {
-      val testGroups = buildGroups(testPoints.map(_.group)).toArray
-      testMatrix.setGroup(testGroups)
+      testMatrix.setGroup(testGroups.result())
     }
 
-    r.setSeed(seed)
-    for (baseMargins <- baseMarginsOpt) {
-      val (trainMargin, testMargin) = baseMargins.partition(_ => r.nextDouble() <= trainTestRatio)
-      trainMatrix.setBaseMargin(trainMargin)
-      testMatrix.setBaseMargin(testMargin)
-    }
+    val trainMargin = fromBaseMarginsToArray(trainBaseMargins.result().iterator)
+    val testMargin = fromBaseMarginsToArray(testBaseMargins.result().iterator)
+    if (trainMargin.isDefined) trainMatrix.setBaseMargin(trainMargin.get)
+    if (testMargin.isDefined) testMatrix.setBaseMargin(testMargin.get)
 
     new Watches(trainMatrix, testMatrix, cacheDirName)
   }
 }
+
+
+/**
+ * Within each RDD partition, group the <code>XGBLabeledPoint</code> by group id.</p>
+ * While loading data into RDD,
+ * the items within a group may be partitioned into the adjacent partitions.
+ * So the fist and last groups within a partition may not be completed
+ * and they need to be skipped from training data.</p>
+ * Make the first and last group as empty list, and they can be filtered out later easily.
+ * @param base collection of <code>XGBLabeledPoint</code>
+ */
+private[spark] class LabeledPointGroupIterator(base: Iterator[XGBLabeledPoint])
+  extends AbstractIterator[Array[XGBLabeledPoint]] {
+  private var firstPointOfNextGroup: XGBLabeledPoint = null
+
+  override def hasNext: Boolean = {
+    return base.hasNext
+  }
+
+  override def next(): Array[XGBLabeledPoint] = {
+    val builder = mutable.ArrayBuilder.make[XGBLabeledPoint]
+    var isFirstGroup = true
+    if (firstPointOfNextGroup != null) {
+      builder += firstPointOfNextGroup
+      isFirstGroup = false
+    }
+
+    var isNewGroup = false
+    var isLastGroup = true
+    while (!isNewGroup && base.hasNext) {
+      val point = base.next()
+      val groupId = if (firstPointOfNextGroup != null) firstPointOfNextGroup.group else point.group
+      firstPointOfNextGroup = point
+      if (point.group == groupId) {
+        builder += point
+      } else {
+        // start a new group
+        isNewGroup = true
+        isLastGroup = false
+      }
+    }
+    val result = builder.result()
+    if (isFirstGroup || isLastGroup) {
```

Comment:
I would suggest at least providing the exact solution (repartition and sortWithinPartition) to the user in this PR (just like exact and approx tree_method in xgb)

what do you think @hcho3 @yanboliang @superbobry

Reply:
@CodingCat to provide multiple options is a good idea (for large training data or smaller ones).

For large training data, dropping a few groups may not affect result, and another major reason is the performance. When the data set is huge, to fix the border of groups in each partition is very expensive, and also there is a memory concern that we should not do too fancy Spark transformations or actions which actually load a lot of data into memory, since we need to save more memory for off-heap usage by Xgboost library.

Reply:
I agree, let’s hands the right of making tradeoff to the user

Reply:
Think of an not too expensive way to stitch the groups back, will update.